### PR TITLE
Add the PCI device type

### PR DESCRIPTION
**Pull request by stgraber**

### CODE REVIEW

**Comment by stgraber** [comment]
Still need to test this with some PCI devices to make sure it's working as expected.

**Comment by stgraber** [comment]
I have now tested this with a couple of PCI devices, it's working as expected.

**Comment by stgraber** [comment]
Should be good to go, jenkins looking happy with it and manual tests looked good too.

File: lxd/device/device_utils_generic.go [hunk 1]
```diff
@@ -16,8 +16,13 @@ func deviceJoinPath(parts ...string) string {
 // validatePCIDevice returns whether a configured PCI device exists. It also returns true, if no device
 // has been specified.
 func validatePCIDevice(config deviceConfig.Device) error {
```
Comment (tomponline) [inline lxd/device/device_utils_generic.go:18]: @stgraber I think that now the `pci` field is not the only field that can store a PCI device address, and that `address` is generic and could be used by other devices for other meaning, it would be better to alter this function to just take a string argument of `pciAddress` and then it would just be a way of avoiding repetitive calls to `!shared.PathExists(fmt.Sprintf("/sys/bus/pci/devices/%s", pciAddress)))`.

This way we are keeping the device specific logic in the per-device structs and not in the generic helper functions.

Then we would update the devices that currently use `validatePCIDevice` to extract their device specific pci address field (and where relevant skip the step if the device specific address field is empty) and pass it into this function. An additional reason for this is that the existing users of this function the `pci` field is optional, whereas this new device requires the field, so we should avoid having the validation environment skip empty fields (although the config validation should have picked up on this anyway).

Also as its a generic PCI helper, it might be good to move this into the `pci` package as `pcidev.DeviceExists()`.

File: lxd/device/pci.go [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+	rules := map[string]func(string) error{
+		"address": validate.IsPCIAddress,
+	}
+
+	err := d.config.Validate(rules)
+	if err != nil {
+		return errors.Wrapf(err, "Failed to validate config")
+	}
+
+	// PCI devices can be specified as "0000:XX:XX.X" or "XX:XX.X".
```
Comment (tomponline) [inline lxd/device/pci.go:35]: @stgraber this pci logic has been repeated a few times already in the code base, I wonder if it would be useful to be move this to the `pci` package as `pcidev.NormaliseAddress()` or `pcidev.CanonicalAddress()`?

File: lxd/device/pci.go [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+	// Get PCI information about the device.
+	pciAddress := d.config["address"]
+	devicePath := filepath.Join("/sys/bus/pci/devices", pciAddress)
+	pciDev, err := pcidev.ParseUeventFile(filepath.Join(devicePath, "uevent"))
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to get PCI device info for %q", pciAddress)
+	}
+
+	saveData["last_state.pci.slot.name"] = pciDev.SlotName
```
Comment (tomponline) [inline lxd/device/pci.go:68]: @stgraber do we need to store slot name in volatile for this device (isn't it always the same as the d.config["address"] value?).

Comment (stgraber) [reply lxd/device/pci.go:68]: We probably don't need to but given we do it everywhere else we deal with PCI devices, might as well keep things consistent :)

Comment (tomponline) [reply lxd/device/pci.go:68]: Fair enough :)

**Comment by stgraber** [comment]
Updated and tested again to confirm it didn't regress:

```
root@amdgpu:~# lxc init images:ubuntu/20.04/cloud test --vm
Creating test
root@amdgpu:~# lxc config device add test gpu1 none
Device gpu1 added to test
root@amdgpu:~# lxc config device add test gpu2 none
Device gpu2 added to test
root@amdgpu:~# lxc config device add test pci1 pci address=26:03.7
Device pci1 added to test
root@amdgpu:~# lxc config device add test pci2 pci address=04:0d.7
Device pci2 added to test
root@amdgpu:~# lxc start test
```

File: lxd/device/pci/pci.go [hunk 4]
```diff
@@ -158,3 +158,32 @@ func deviceProbeWait(pciDev Device) error {
 
 	return fmt.Errorf("Device took too long to activate at %q", driverPath)
 }
+
+// NormaliseAddress converts common PCI address notation to the kernel's notation.
+func NormaliseAddress(addr string) string {
+	// Ignore empty values.
+	if addr == "" {
```
Comment (tomponline) [inline lxd/device/pci/pci.go:165]: @stgraber I would prefer this function to not have device specific logic in it (as the new pci device wouldn't accept an empty device string).

File: lxd/device/pci/pci.go [hunk 5]
```diff
@@ -158,3 +158,32 @@ func deviceProbeWait(pciDev Device) error {
...
+
+	// Ensure all addresses are lowercase.
+	addr = strings.ToLower(addr)
+
+	return addr
+}
+
+// DeviceExists checks whether a configured PCI device exists.
+func DeviceExists(addr string) error {
+	if addr != "" && !shared.PathExists(fmt.Sprintf("/sys/bus/pci/devices/%s", addr)) {
```
Comment (tomponline) [inline lxd/device/pci/pci.go:184]: @stgraber same with this one, the empty check is device specific and seems like it should be in each device (depending on whether it is optional or not).

Comment (stgraber) [reply lxd/device/pci/pci.go:184]: For that one it's making `DeviceExists` a one-line function and adding 4 lines of boiler plate to all callers so that doesn't seem worth it.

Instead I'll revert to having a `validatePCIDevice` in device_utils_generic.go as we did before and have that take the address directly.

Comment (tomponline) [reply lxd/device/pci/pci.go]: I suppose either way it will return true on empty, as `/sys/bus/pci/devices/` will classify as existing.

**Review by tomponline — APPROVED** [review]
